## 4.1 Overview of a DataFrame

DataFrame은 행과 열이 있는 2차원 data table이다.

Series와 마찬가지로 index label과 index 위치를 각 행에 할당한다. 또한 각 열에 label과 위치를 할당한다.

DataFrame은 공통 index(=행 번호)를 갖는 Series object의 collection라고 생각하면 쉽게 이해할 수 있다.

In [1]:
import pandas as pd
import numpy as np

### 4.1.1 Creating DataFrame from Dictionary

In [2]:
#dictionary의 key값은 DataFrame의 각 column 속성의 이름이 되고,
#value는 DataFrame의 값이 된다.
city_data = {
    "City": ["New York City", "Paris", "Barcelona", "Rome"],
    "Country": ["United States", "France", "Spain", "Italy"],
    "Population": pd.Series([8600000, 2141000, 5515000, 2873000])
}

cities = pd.DataFrame(city_data)
cities

,City,Country,Population
0,New York City,United States,8600000
1,Paris,France,2141000
2,Barcelona,Spain,5515000
3,Rome,Italy,2873000


In [3]:
# The two lines below are equivalent
#column headers을 index label로 바꾼다.
cities.transpose()
cities.T

,0,1,2,3
City,New York City,Paris,Barcelona,Rome
Country,United States,France,Spain,Italy
Population,8600000,2141000,5515000,2873000


### 4.1.2 Creating a DataFrame from a NumPy ndarray

In [4]:
random_data = np.random.randint(1, 101, [3, 5])
random_data

array([[66,  5, 44, 93, 45],
       [27,  2, 94, 72,  2],
       [42, 31, 98, 95, 89]])

In [5]:
#ndarray을 값으로 갖는 DataFrame을 만들 수도 있다.
pd.DataFrame(data = random_data)

,0,1,2,3,4
0,66,5,44,93,45
1,27,2,94,72,2
2,42,31,98,95,89


In [6]:
#index 매개변수에 list, tuple, ndarray와 같은 반복 가능한 object를 넘길 수 있으며
#이 매개변수를 통해 원하는 행 label을 설정할 수 있다.
#주의할 점은 label로 할 object의 길이는 DataFrame의 행 개수와 동일해야 한다.

row_labels = ["Morning", "Afternoon", "Evening"]
temperatures = pd.DataFrame(
    data = random_data, index = row_labels
)
temperatures

,0,1,2,3,4
Morning,66,5,44,93,45
Afternoon,27,2,94,72,2
Evening,42,31,98,95,89


In [7]:
#columns 매개변수를 사용하여 열 이름을 설정할 수 있다.

row_labels = ["Morning", "Afternoon", "Evening"]
column_lables = (
    "Monday",
    "Tuesday",
    "Wednesday",
    "Thursday",
    "Friday"
)

pd.DataFrame(
    data = random_data,
    index = row_labels,
    columns = column_lables,
)

,Monday,Tuesday,Wednesday,Thursday,Friday
Morning,66,5,44,93,45
Afternoon,27,2,94,72,2
Evening,42,31,98,95,89


In [8]:
#pandas는 중복된 행과 열 index를 허용한다.

row_labels = ["Morning", "Afternoon", "Morning"]
column_lables = [
    "Monday",
    "Tuesday",
    "Wednesday",
    "Tuesday",
    "Friday"
]

pd.DataFrame(
    data = random_data,
    index = row_labels,
    columns = column_lables
)


,Monday,Tuesday,Wednesday,Tuesday,Friday
Morning,66,5,44,93,45
Afternoon,27,2,94,72,2
Morning,42,31,98,95,89


## 4.2 Similarities between Series and DataFrames

### 4.2.1 Importing a DataFrame with the read_csv Function

In [9]:
pd.read_csv("nba.csv")

,Name,Team,Position,Birthday,Salary
0,Shake Milton,Philadelphia 76ers,SG,9/26/96,1445697
1,Christian Wood,Detroit Pistons,PF,9/27/95,1645357
2,PJ Washington,Charlotte Hornets,PF,8/23/98,3831840
3,Derrick Rose,Detroit Pistons,PG,10/4/88,7317074
4,Marial Shayok,Philadelphia 76ers,G,7/26/95,79568
...,...,...,...,...,...
445,Austin Rivers,Houston Rockets,PG,8/1/92,2174310
446,Harry Giles,Sacramento Kings,PF,4/22/98,2578800
447,Robin Lopez,Milwaukee Bucks,C,4/1/88,4767000
448,Collin Sexton,Cleveland Cavaliers,PG,1/4/99,4764960


In [10]:
pd.read_csv("nba.csv", parse_dates = ["Birthday"])

/var/folders/g2/m_rjbg_d4fq330k9g6z2v24m0000gn/T/ipykernel_1034/3048234929.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  pd.read_csv("nba.csv", parse_dates = ["Birthday"])


,Name,Team,Position,Birthday,Salary
0,Shake Milton,Philadelphia 76ers,SG,1996-09-26,1445697
1,Christian Wood,Detroit Pistons,PF,1995-09-27,1645357
2,PJ Washington,Charlotte Hornets,PF,1998-08-23,3831840
3,Derrick Rose,Detroit Pistons,PG,1988-10-04,7317074
4,Marial Shayok,Philadelphia 76ers,G,1995-07-26,79568
...,...,...,...,...,...
445,Austin Rivers,Houston Rockets,PG,1992-08-01,2174310
446,Harry Giles,Sacramento Kings,PF,1998-04-22,2578800
447,Robin Lopez,Milwaukee Bucks,C,1988-04-01,4767000
448,Collin Sexton,Cleveland Cavaliers,PG,1999-01-04,4764960


In [11]:
nba = pd.read_csv("nba.csv", parse_dates = ["Birthday"])

/var/folders/g2/m_rjbg_d4fq330k9g6z2v24m0000gn/T/ipykernel_1034/2125825459.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  nba = pd.read_csv("nba.csv", parse_dates = ["Birthday"])


### 4.2.2 Shared and Exclusive Attributes between Series and DataFrames

In [12]:
pd.Series([1, 2, 3]).dtype

dtype('int64')

In [13]:
#밑의 출력처럼 DataFrame은 각 열마다 다른 유형의 데이터를 담을 수 있다. 
#이때 dtypes의 반환형은 Series타입이다.
print(type(nba.dtypes))
nba.dtypes

<class 'pandas.core.series.Series'>


Name                object
Team                object
Position            object
Birthday    datetime64[ns]
Salary               int64
dtype: object

In [14]:
#열의 데이터 타입의 빈도수를 다음과 같이 계산할 수 있다.
#왜냐, dtypes의 반환형이 Series이기 때문이다.
nba.dtypes.value_counts()

object            3
datetime64[ns]    1
int64             1
Name: count, dtype: int64

In [15]:
#DataFrame의 index를 나타낸다. 
nba.index

RangeIndex(start=0, stop=450, step=1)

In [16]:
#DataFrame의 column index에 접근할 수 있다.
nba.columns

Index(['Name', 'Team', 'Position', 'Birthday', 'Salary'], dtype='object')

In [17]:
#pandas object의 차원 수를 반환한다. DataFrame은 2차원임을 알 수 있다.
nba.ndim

2

In [18]:
#이 DataFrame은 450개의 행과 5개의 열로 이뤄져있음을 알 수 있다.
nba.shape

(450, 5)

In [19]:
#dataset에 있는 값의 전체 개수를 반환한다. 이때 NaN과 같은 결측값도 개수에 포함된다.
nba.size

2250

In [20]:
#결측값을 제외한 각 열에 존재하는 값의 개수
nba.count()

Name        450
Team        450
Position    450
Birthday    450
Salary      450
dtype: int64

In [21]:
nba.count().sum()

np.int64(2250)

In [22]:
#결측값이 존재하는 DataFrame생성
data = {
    "A": [1, np.nan],
    "B": [2, 3]
}

df = pd.DataFrame(data)
df

,A,B
0,1.0,2
1,NaN,3


In [23]:
df.size

4

In [24]:
df.count()

A    1
B    2
dtype: int64

In [25]:
df.count().sum()

np.int64(3)

### 4.2.3 Shared Methods between Series and DataFrames

In [26]:
nba.head(2)

,Name,Team,Position,Birthday,Salary
0,Shake Milton,Philadelphia 76ers,SG,1996-09-26,1445697
1,Christian Wood,Detroit Pistons,PF,1995-09-27,1645357


In [27]:
nba.tail(n = 3)

,Name,Team,Position,Birthday,Salary
447,Robin Lopez,Milwaukee Bucks,C,1988-04-01,4767000
448,Collin Sexton,Cleveland Cavaliers,PG,1999-01-04,4764960
449,Ricky Rubio,Phoenix Suns,PG,1990-10-21,16200000


In [28]:
nba.tail()

,Name,Team,Position,Birthday,Salary
445,Austin Rivers,Houston Rockets,PG,1992-08-01,2174310
446,Harry Giles,Sacramento Kings,PF,1998-04-22,2578800
447,Robin Lopez,Milwaukee Bucks,C,1988-04-01,4767000
448,Collin Sexton,Cleveland Cavaliers,PG,1999-01-04,4764960
449,Ricky Rubio,Phoenix Suns,PG,1990-10-21,16200000


In [29]:
#DataFrame에서 임의의 행을 추출한다(랜덤하게 추출).
#이 method의 매개변수는 반환할 행의 개수이다.
nba.sample(3)

,Name,Team,Position,Birthday,Salary
13,Tyler Johnson,Phoenix Suns,PG,1992-05-07,19245370
422,Jonah Bolden,Philadelphia 76ers,PF,1996-01-02,1698450
363,Daryl Macon,Miami Heat,SG,1995-11-29,79568


In [30]:
#각 열별로 고유한 값의 개수를 나타내는 Series를 반환한다.
nba.nunique()

Name        450
Team         30
Position      9
Birthday    430
Salary      269
dtype: int64

        Name        450
        Team         30
        Position      9
        Birthday    430
        Salary      269

30개의 고유 팀, 269개의 고유한 연봉, 9개의 고유한 포지션이 있다는 것을 알 수 있다.


In [31]:
#각 열에 있는 최대값을 나타내는 Series를 반환한다.
#이때 날짜/시간 열의 최댓값은 시간 순서로 가장 최근 날짜이다.
nba.max()

Name             Zylan Cheatham
Team         Washington Wizards
Position                     SG
Birthday    2000-12-23 00:00:00
Salary                 40231758
dtype: object

In [32]:
#각 열의 최솟값을 나타내는 Series를 반환한다.
nba.min()

Name               Aaron Gordon
Team              Atlanta Hawks
Position                      C
Birthday    1977-01-26 00:00:00
Salary                    79568
dtype: object

In [33]:
#Salary column에서 가장 높은 값을 갖는 4개의 행을 추출
nba.nlargest(n = 4, columns = "Salary")

,Name,Team,Position,Birthday,Salary
205,Stephen Curry,Golden State Warriors,PG,1988-03-14,40231758
38,Chris Paul,Oklahoma City Thunder,PG,1985-05-06,38506482
219,Russell Westbrook,Houston Rockets,PG,1988-11-12,38506482
251,John Wall,Washington Wizards,PG,1990-09-06,38199000


In [34]:
#Birthday column에서 가장 오래된 날짜 3개를 추출한다.
nba.nsmallest(n = 3, columns = ["Birthday"])

,Name,Team,Position,Birthday,Salary
98,Vince Carter,Atlanta Hawks,PF,1977-01-26,2564753
196,Udonis Haslem,Miami Heat,C,1980-06-09,2564753
262,Kyle Korver,Milwaukee Bucks,PF,1981-03-17,6004753


nlargest()와 nsmallest()는 숫자 또는 날짜/시간 열에서만 작동된다.

In [35]:
#각 열에 있는 값을 더할 때, 숫자를 합한 결과만 보고 싶다면 numeric_only 매개변수을 True로 설정한다.
nba.sum(numeric_only = True)

Salary    3444112694
dtype: int64

In [36]:
nba.mean(numeric_only = True)

Salary    7.653584e+06
dtype: float64

In [37]:
#중앙값
print(type(nba.median(numeric_only = True)))
nba.median(numeric_only = True)

<class 'pandas.core.series.Series'>


Salary    3303074.5
dtype: float64

In [38]:
#최빈값
print(type(nba.mode(numeric_only=True)))
nba.mode(numeric_only = True)

<class 'pandas.core.frame.DataFrame'>


,Salary
0,79568


In [39]:
#표준편차
nba.std(numeric_only = True)

Salary    9.288810e+06
dtype: float64

## 4.3 Sorting a DataFrame

### 4.3.1 Sorting by Single Column

In [51]:
# The two lines below are equivalent
#Name column을 알바벳 기준으로 오름차순으로 정렬하겠다.
nba.sort_values("Name")
nba.sort_values(by = ["Name"])

,Name,Team,Position,Birthday,Salary
52,Aaron Gordon,Orlando Magic,PF,1995-09-16,19863636
101,Aaron Holiday,Indiana Pacers,PG,1996-09-30,2239200
437,Abdel Nader,Oklahoma City Thunder,SF,1993-09-25,1618520
81,Adam Mokoka,Chicago Bulls,G,1998-07-18,79568
399,Admiral Schofield,Washington Wizards,SF,1997-03-30,1000000
...,...,...,...,...,...
159,Zach LaVine,Chicago Bulls,PG,1995-03-10,19500000
302,Zach Norvell,Los Angeles Lakers,SG,1997-12-09,79568
312,Zhaire Smith,Philadelphia 76ers,SG,1999-06-04,3058800
137,Zion Williamson,New Orleans Pelicans,F,2000-07-06,9757440


In [57]:
nba.sort_values("Name", ascending = False).head()

,Name,Team,Position,Birthday,Salary
248,Zylan Cheatham,New Orleans Pelicans,SF,1995-11-17,79568
137,Zion Williamson,New Orleans Pelicans,F,2000-07-06,9757440
312,Zhaire Smith,Philadelphia 76ers,SG,1999-06-04,3058800
302,Zach Norvell,Los Angeles Lakers,SG,1997-12-09,79568
159,Zach LaVine,Chicago Bulls,PG,1995-03-10,19500000


In [58]:
nba.sort_values("Birthday", ascending = False).head()

,Name,Team,Position,Birthday,Salary
136,Sekou Doumbouya,Detroit Pistons,SF,2000-12-23,3285120
432,Talen Horton-Tucker,Los Angeles Lakers,GF,2000-11-25,898310
137,Zion Williamson,New Orleans Pelicans,F,2000-07-06,9757440
313,RJ Barrett,New York Knicks,SG,2000-06-14,7839960
392,Jalen Lecque,Phoenix Suns,G,2000-06-13,898310


### 4.3.2 Sorting by Multiple Columns

In [61]:
#by매개변수에 list를 전달하여 여러 column을 정렬할 수 있다.
#Team을 기준으로 정렬한 다음 이름을 기준으로 정렬한다.
nba.sort_values(by = ["Team", "Name"])

,Name,Team,Position,Birthday,Salary
359,Alex Len,Atlanta Hawks,C,1993-06-16,4160000
167,Allen Crabbe,Atlanta Hawks,SG,1992-04-09,18500000
276,Brandon Goodwin,Atlanta Hawks,PG,1995-10-02,79568
438,Bruno Fernando,Atlanta Hawks,C,1998-08-15,1400000
194,Cam Reddish,Atlanta Hawks,SF,1999-09-01,4245720
...,...,...,...,...,...
418,Jordan McRae,Washington Wizards,PG,1991-03-28,1645357
273,Justin Robinson,Washington Wizards,PG,1997-10-12,898310
428,Moritz Wagner,Washington Wizards,C,1997-04-26,2063520
21,Rui Hachimura,Washington Wizards,PF,1998-02-08,4469160


In [63]:
nba.sort_values(by=["Team","Birthday"],ascending=False)

,Name,Team,Position,Birthday,Salary
80,Isaac Bonga,Washington Wizards,PG,1999-11-08,1416852
21,Rui Hachimura,Washington Wizards,PF,1998-02-08,4469160
273,Justin Robinson,Washington Wizards,PG,1997-10-12,898310
36,Thomas Bryant,Washington Wizards,C,1997-07-31,8000000
428,Moritz Wagner,Washington Wizards,C,1997-04-26,2063520
...,...,...,...,...,...
359,Alex Len,Atlanta Hawks,C,1993-06-16,4160000
167,Allen Crabbe,Atlanta Hawks,SG,1992-04-09,18500000
28,Evan Turner,Atlanta Hawks,PG,1988-10-27,18606556
111,Chandler Parsons,Atlanta Hawks,SF,1988-10-25,25102512


In [64]:
nba.sort_values(["Team", "Name"], ascending = False)

,Name,Team,Position,Birthday,Salary
36,Thomas Bryant,Washington Wizards,C,1997-07-31,8000000
21,Rui Hachimura,Washington Wizards,PF,1998-02-08,4469160
428,Moritz Wagner,Washington Wizards,C,1997-04-26,2063520
273,Justin Robinson,Washington Wizards,PG,1997-10-12,898310
418,Jordan McRae,Washington Wizards,PG,1991-03-28,1645357
...,...,...,...,...,...
194,Cam Reddish,Atlanta Hawks,SF,1999-09-01,4245720
438,Bruno Fernando,Atlanta Hawks,C,1998-08-15,1400000
276,Brandon Goodwin,Atlanta Hawks,PG,1995-10-02,79568
167,Allen Crabbe,Atlanta Hawks,SG,1992-04-09,18500000


In [65]:
#Team은 오름차순으로 정렬하고, Salary는 내림차순으로 정렬하기 위해, ascending에 bool list를 전달하였다.
nba.sort_values(
    by = ["Team", "Salary"], ascending = [True, False]
)

,Name,Team,Position,Birthday,Salary
111,Chandler Parsons,Atlanta Hawks,SF,1988-10-25,25102512
28,Evan Turner,Atlanta Hawks,PG,1988-10-27,18606556
167,Allen Crabbe,Atlanta Hawks,SG,1992-04-09,18500000
213,De'Andre Hunter,Atlanta Hawks,SF,1997-12-02,7068360
339,Jabari Parker,Atlanta Hawks,PF,1995-03-15,6500000
...,...,...,...,...,...
80,Isaac Bonga,Washington Wizards,PG,1999-11-08,1416852
399,Admiral Schofield,Washington Wizards,SF,1997-03-30,1000000
273,Justin Robinson,Washington Wizards,PG,1997-10-12,898310
283,Garrison Mathews,Washington Wizards,SG,1996-10-24,79568


In [66]:
nba = nba.sort_values(
    by = ["Team", "Salary"],
    ascending = [True, False]
)

## 4.4 Sorting by Index

In [67]:
nba.head()

,Name,Team,Position,Birthday,Salary
111,Chandler Parsons,Atlanta Hawks,SF,1988-10-25,25102512
28,Evan Turner,Atlanta Hawks,PG,1988-10-27,18606556
167,Allen Crabbe,Atlanta Hawks,SG,1992-04-09,18500000
213,De'Andre Hunter,Atlanta Hawks,SF,1997-12-02,7068360
339,Jabari Parker,Atlanta Hawks,PF,1995-03-15,6500000


### 4.4.1 Sorting by Row Index

In [68]:
# The two lines below are equivalent
#index위치를 기준으로 정렬한다.
nba.sort_index().head()
nba.sort_index(ascending = True).head()

,Name,Team,Position,Birthday,Salary
0,Shake Milton,Philadelphia 76ers,SG,1996-09-26,1445697
1,Christian Wood,Detroit Pistons,PF,1995-09-27,1645357
2,PJ Washington,Charlotte Hornets,PF,1998-08-23,3831840
3,Derrick Rose,Detroit Pistons,PG,1988-10-04,7317074
4,Marial Shayok,Philadelphia 76ers,G,1995-07-26,79568


In [69]:
nba.sort_index(ascending = False).head()

,Name,Team,Position,Birthday,Salary
449,Ricky Rubio,Phoenix Suns,PG,1990-10-21,16200000
448,Collin Sexton,Cleveland Cavaliers,PG,1999-01-04,4764960
447,Robin Lopez,Milwaukee Bucks,C,1988-04-01,4767000
446,Harry Giles,Sacramento Kings,PF,1998-04-22,2578800
445,Austin Rivers,Houston Rockets,PG,1992-08-01,2174310


In [70]:
nba = nba.sort_index()

### 4.4.2 Sorting by Column Index

In [79]:
# The two lines below are equivalent
#열의 이름을 순서대로 정렬하고자 할 때 axis=1혹은 axis="columns"을 사용하면 된다(행 순서대로 정렬은 axis=0, axis="rows"다.).
#그럼 출력과 같이 알파벳 순서가 가장 빠른 Birthday가 먼저오고, 그 다음 Name이런식으로 오게 된다.
nba.sort_index(axis = "columns").head()
nba.sort_index(axis = 1).head()

,Birthday,Name,Position,Salary,Team
0,1996-09-26,Shake Milton,SG,1445697,Philadelphia 76ers
1,1995-09-27,Christian Wood,PF,1645357,Detroit Pistons
2,1998-08-23,PJ Washington,PF,3831840,Charlotte Hornets
3,1988-10-04,Derrick Rose,PG,7317074,Detroit Pistons
4,1995-07-26,Marial Shayok,G,79568,Philadelphia 76ers


In [80]:
#내림차순으로 정렬하고 싶을 때
nba.sort_index(axis = "columns", ascending = False).head()

,Team,Salary,Position,Name,Birthday
0,Philadelphia 76ers,1445697,SG,Shake Milton,1996-09-26
1,Detroit Pistons,1645357,PF,Christian Wood,1995-09-27
2,Charlotte Hornets,3831840,PF,PJ Washington,1998-08-23
3,Detroit Pistons,7317074,PG,Derrick Rose,1988-10-04
4,Philadelphia 76ers,79568,G,Marial Shayok,1995-07-26


## 4.5 Setting a New Index

In [81]:
# The two lines below are equivalent
#set_index() method는 지정된 열이 index label로 설정된 새 DataFrame을 반환한다.
#이때는 Name이 index label로 설정된다.
nba.set_index(keys = "Name")
nba.set_index("Name")

,Team,Position,Birthday,Salary
Name,,,,
Shake Milton,Philadelphia 76ers,SG,1996-09-26,1445697
Christian Wood,Detroit Pistons,PF,1995-09-27,1645357
PJ Washington,Charlotte Hornets,PF,1998-08-23,3831840
Derrick Rose,Detroit Pistons,PG,1988-10-04,7317074
Marial Shayok,Philadelphia 76ers,G,1995-07-26,79568
...,...,...,...,...
Austin Rivers,Houston Rockets,PG,1992-08-01,2174310
Harry Giles,Sacramento Kings,PF,1998-04-22,2578800
Robin Lopez,Milwaukee Bucks,C,1988-04-01,4767000


In [82]:
nba = nba.set_index(keys = "Name")

In [85]:
#dataset을 가져올 때 바로 index label을 설정할 수도 있다.
nba = pd.read_csv(
    "nba.csv", parse_dates = ["Birthday"], index_col = "Name"
)
nba

/var/folders/g2/m_rjbg_d4fq330k9g6z2v24m0000gn/T/ipykernel_1034/335948394.py:2: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  nba = pd.read_csv(


,Team,Position,Birthday,Salary
Name,,,,
Shake Milton,Philadelphia 76ers,SG,1996-09-26,1445697
Christian Wood,Detroit Pistons,PF,1995-09-27,1645357
PJ Washington,Charlotte Hornets,PF,1998-08-23,3831840
Derrick Rose,Detroit Pistons,PG,1988-10-04,7317074
Marial Shayok,Philadelphia 76ers,G,1995-07-26,79568
...,...,...,...,...
Austin Rivers,Houston Rockets,PG,1992-08-01,2174310
Harry Giles,Sacramento Kings,PF,1998-04-22,2578800
Robin Lopez,Milwaukee Bucks,C,1988-04-01,4767000


## 4.6 Selecting Columns and Rows from a DataFrame

### 4.6.1 Selecting a Single Column from a DataFrame

In [86]:
#DataFrame내 각 Series column은 객체 속성으로 접근하여 그것만 추출할 수 있다.
#다음은 Salary column을 추출한 결과이다.
nba.Salary

Name
Shake Milton       1445697
Christian Wood     1645357
PJ Washington      3831840
Derrick Rose       7317074
Marial Shayok        79568
                    ...   
Austin Rivers      2174310
Harry Giles        2578800
Robin Lopez        4767000
Collin Sexton      4764960
Ricky Rubio       16200000
Name: Salary, Length: 450, dtype: int64

In [88]:
#다음은 Position column을 추출한 결과이다.
#이렇게 []대괄호를 사용하는 방법이 더 권장된다.
nba["Position"]

Name
Shake Milton      SG
Christian Wood    PF
PJ Washington     PF
Derrick Rose      PG
Marial Shayok      G
                  ..
Austin Rivers     PG
Harry Giles       PF
Robin Lopez        C
Collin Sexton     PG
Ricky Rubio       PG
Name: Position, Length: 450, dtype: object

In [90]:
type(nba["Position"])

pandas.core.series.Series

### 4.6.2 Selecting Multiple Columns from a DataFrame

In [89]:
#여러 개의 Series column을 추출하고 싶을 때 다음과 같이 대괄호를 두 번 감싸서 사용하면 된다.
nba[["Salary", "Birthday"]].head()

,Salary,Birthday
Name,,
Shake Milton,1445697,1996-09-26
Christian Wood,1645357,1995-09-27
PJ Washington,3831840,1998-08-23
Derrick Rose,7317074,1988-10-04
Marial Shayok,79568,1995-07-26


In [95]:
type(nba[["Position","Salary"]])

pandas.core.frame.DataFrame

In [96]:
nba[["Birthday", "Salary"]].head()

,Birthday,Salary
Name,,
Shake Milton,1996-09-26,1445697
Christian Wood,1995-09-27,1645357
PJ Washington,1998-08-23,3831840
Derrick Rose,1988-10-04,7317074
Marial Shayok,1995-07-26,79568


In [101]:
nba.dtypes

Team                object
Position            object
Birthday    datetime64[ns]
Salary               int64
dtype: object

In [97]:
#select_dtypes() method을 사용하여 data type에 따라 column을 선택할 수도 있다.
#이때는 object에 대한 열만 선택한다.
nba.select_dtypes(include = "object")

,Team,Position
Name,,
Shake Milton,Philadelphia 76ers,SG
Christian Wood,Detroit Pistons,PF
PJ Washington,Charlotte Hornets,PF
Derrick Rose,Detroit Pistons,PG
Marial Shayok,Philadelphia 76ers,G
...,...,...
Austin Rivers,Houston Rockets,PG
Harry Giles,Sacramento Kings,PF
Robin Lopez,Milwaukee Bucks,C


In [103]:
nba.select_dtypes(include=["object","datetime"])

,Team,Position,Birthday
Name,,,
Shake Milton,Philadelphia 76ers,SG,1996-09-26
Christian Wood,Detroit Pistons,PF,1995-09-27
PJ Washington,Charlotte Hornets,PF,1998-08-23
Derrick Rose,Detroit Pistons,PG,1988-10-04
Marial Shayok,Philadelphia 76ers,G,1995-07-26
...,...,...,...
Austin Rivers,Houston Rockets,PG,1992-08-01
Harry Giles,Sacramento Kings,PF,1998-04-22
Robin Lopez,Milwaukee Bucks,C,1988-04-01


In [104]:
#object에 대한 열과 정수형에 대한 열을 제외한 모든 열을 선택한다.
nba.select_dtypes(exclude = ["object", "int"])

,Birthday
Name,
Shake Milton,1996-09-26
Christian Wood,1995-09-27
PJ Washington,1998-08-23
Derrick Rose,1988-10-04
Marial Shayok,1995-07-26
...,...
Austin Rivers,1992-08-01
Harry Giles,1998-04-22
Robin Lopez,1988-04-01


## 4.7 Selecting Rows from a DataFrame

### 4.7.1 Extracting Rows by Index Label

In [105]:
#loc는 index label로 데이터를 추출한다.
#nba의 index label은 Name이므로 사람 이름으로 데이터를 추출할 수 있다.
nba.loc["LeBron James"]

Team         Los Angeles Lakers
Position                     PF
Birthday    1984-12-30 00:00:00
Salary                 37436858
Name: LeBron James, dtype: object

In [106]:
#여러 행을 추출
nba.loc[["Kawhi Leonard", "Paul George"]]

,Team,Position,Birthday,Salary
Name,,,,
Kawhi Leonard,Los Angeles Clippers,SF,1991-06-29,32742000
Paul George,Los Angeles Clippers,SF,1990-05-02,33005556


In [111]:
nba.loc[["Paul George", "Kawhi Leonard"]]

,Team,Position,Birthday,Salary
Name,,,,
Paul George,Los Angeles Clippers,SF,1990-05-02,33005556
Kawhi Leonard,Los Angeles Clippers,SF,1991-06-29,32742000


In [112]:
#Otta Porter과 Patrick Beverley 사이에 있는 모든 선수를 추출할 수 있다.
nba.sort_index().loc["Otto Porter":"Patrick Beverley"]

,Team,Position,Birthday,Salary
Name,,,,
Otto Porter,Chicago Bulls,SF,1993-06-03,27250576
PJ Dozier,Denver Nuggets,PG,1996-10-25,79568
PJ Washington,Charlotte Hornets,PF,1998-08-23,3831840
Pascal Siakam,Toronto Raptors,PF,1994-04-02,2351838
Pat Connaughton,Milwaukee Bucks,SG,1993-01-06,1723050
Patrick Beverley,Los Angeles Clippers,PG,1988-07-12,12345680


In [113]:
players = ["Otto Porter", "PJ Dozier", "PJ Washington"]
players[0:2]

['Otto Porter', 'PJ Dozier']

In [127]:
#Zach Collins부터 끝까지 간격은 2칸씩
nba.sort_index().loc["Zach Collins"::2]

,Team,Position,Birthday,Salary
Name,,,,
Zach Collins,Portland Trail Blazers,C,1997-11-19,4240200
Zach Norvell,Los Angeles Lakers,SG,1997-12-09,79568
Zion Williamson,New Orleans Pelicans,F,2000-07-06,9757440


In [120]:
#첫 행부터 "Al Horford"까지 범위에 있는 모든 선수를 반환한다.
nba.sort_index().loc[:"Al Horford"]

,Team,Position,Birthday,Salary
Name,,,,
Aaron Gordon,Orlando Magic,PF,1995-09-16,19863636
Aaron Holiday,Indiana Pacers,PG,1996-09-30,2239200
Abdel Nader,Oklahoma City Thunder,SF,1993-09-25,1618520
Adam Mokoka,Chicago Bulls,G,1998-07-18,79568
Admiral Schofield,Washington Wizards,SF,1997-03-30,1000000
Al Horford,Philadelphia 76ers,C,1986-06-03,28000000


**NOTE**: I've commented out the code below so that the Notebook can run without raising an error.

In [122]:
#index label이 DataFrame에 존재하지 않으면 예외를 발생시킨다.
nba.loc["Bugs Bunny"]

KeyError: 'Bugs Bunny'

### 4.7.2 Extracting Rows by Index Position

In [123]:
#iloc는 index 위치별로 행을 추출한다.
#300번째 행을 추출한다.
nba.iloc[300]

Team             Denver Nuggets
Position                     PF
Birthday    1999-04-03 00:00:00
Salary                  1416852
Name: Jarred Vanderbilt, dtype: object

In [124]:
#100번째, 200번째, 300번째, 400번째 행을 추출한다.
nba.iloc[[100, 200, 300, 400]]

,Team,Position,Birthday,Salary
Name,,,,
Brian Bowen,Indiana Pacers,SG,1998-10-02,79568
Marco Belinelli,San Antonio Spurs,SF,1986-03-25,5846154
Jarred Vanderbilt,Denver Nuggets,PF,1999-04-03,1416852
Louis King,Detroit Pistons,F,1999-04-06,79568


In [125]:
#400번째부터 404번째까지 행을 추출한다.
nba.iloc[400:404]

,Team,Position,Birthday,Salary
Name,,,,
Louis King,Detroit Pistons,F,1999-04-06,79568
Kostas Antetokounmpo,Los Angeles Lakers,PF,1997-11-20,79568
Rodions Kurucs,Brooklyn Nets,PF,1998-02-05,1699236
Spencer Dinwiddie,Brooklyn Nets,PG,1993-04-06,10605600


In [130]:
#400번째부터 끝까지 5칸 간격으로 행을 추출한다.
nba.iloc[400::5]

,Team,Position,Birthday,Salary
Name,,,,
Louis King,Detroit Pistons,F,1999-04-06,79568
Romeo Langford,Boston Celtics,SG,1999-10-25,3458400
Ty Jerome,Phoenix Suns,G,1997-07-08,2193480
Andre Iguodala,Memphis Grizzlies,SF,1984-01-28,17185185
John Henson,Cleveland Cavaliers,C,1990-12-28,9732396
Semi Ojeleye,Boston Celtics,PF,1994-12-05,1618520
Danuel House,Houston Rockets,SF,1993-06-07,3540000
Kadeem Allen,New York Knicks,PG,1993-01-15,79568
Jared Dudley,Los Angeles Lakers,PF,1985-07-10,2564753


In [131]:
nba.iloc[:2]

,Team,Position,Birthday,Salary
Name,,,,
Shake Milton,Philadelphia 76ers,SG,1996-09-26,1445697
Christian Wood,Detroit Pistons,PF,1995-09-27,1645357


In [132]:
nba.iloc[447:]

,Team,Position,Birthday,Salary
Name,,,,
Robin Lopez,Milwaukee Bucks,C,1988-04-01,4767000
Collin Sexton,Cleveland Cavaliers,PG,1999-01-04,4764960
Ricky Rubio,Phoenix Suns,PG,1990-10-21,16200000


In [133]:
#마지막에서 10번재 행부터 마지막에서 6번째 행까지 추출한다.
nba.iloc[-10:-6]

,Team,Position,Birthday,Salary
Name,,,,
Jared Dudley,Los Angeles Lakers,PF,1985-07-10,2564753
Max Strus,Chicago Bulls,SG,1996-03-28,79568
Kevon Looney,Golden State Warriors,C,1996-02-06,4464286
Willy Hernangomez,Charlotte Hornets,C,1994-05-27,1557250


In [134]:
nba.iloc[0:10:2]

,Team,Position,Birthday,Salary
Name,,,,
Shake Milton,Philadelphia 76ers,SG,1996-09-26,1445697
PJ Washington,Charlotte Hornets,PF,1998-08-23,3831840
Marial Shayok,Philadelphia 76ers,G,1995-07-26,79568
Kendrick Nunn,Miami Heat,SG,1995-08-03,1416852
Brook Lopez,Milwaukee Bucks,C,1988-04-01,12093024


### 4.7.3 Extracting Values from Specific Columns

loc과 iloc 속성은 두 번재 인수로 추출할 열을 받는다.

In [136]:
#Giannis Antetokounmpo에 대한 행 중에서 Team열에 있는 값을 가져온다.
nba.loc["Giannis Antetokounmpo", "Team"]

'Milwaukee Bucks'

In [137]:
#James Harden에 대한 행 중에서 Position열과 Birthday열에 있는 값을 가져온다.
nba.loc["James Harden", ["Position", "Birthday"]]

Position                     PG
Birthday    1989-08-26 00:00:00
Name: James Harden, dtype: object

In [138]:
#여러 개의 행에서 그 각각의 Team값과 Salary값을 추출한다.
nba.loc[
    ["Russell Westbrook", "Anthony Davis"],
    ["Team", "Salary"]
]

,Team,Salary
Name,,
Russell Westbrook,Houston Rockets,38506482
Anthony Davis,Los Angeles Lakers,27093019


In [140]:
#Joel Embiid에 대한 행 중에서 Position열부터 Salary열까지의 열 값을 추출한다.
#이때 이러한 슬라이드 구문을 사용할 때는 DataFrame에서 보여지는 순서에 맞춰 열의 이름을 전달해야 한다.
nba.loc["Joel Embiid", "Position":"Salary"]

Position                      C
Birthday    1994-03-16 00:00:00
Salary                 27504630
Name: Joel Embiid, dtype: object

In [141]:
#Salary열이 Position열보다 뒤에 있지만 Salary를 먼저 입력했을 때 다음과 같은 빈 결과가 반환된다.
nba.loc["Joel Embiid", "Salary":"Position"]

Series([], Name: Joel Embiid, dtype: object)

In [142]:
#index 57에 있는 행과 index 3에 있는 열(Salary)의 교차점에 있는 값을 추출한다.
nba.iloc[57, 3]

np.int64(796806)

In [143]:
nba.iloc[100:104, :3]

,Team,Position,Birthday
Name,,,
Brian Bowen,Indiana Pacers,SG,1998-10-02
Aaron Holiday,Indiana Pacers,PG,1996-09-30
Troy Daniels,Los Angeles Lakers,SG,1991-07-15
Buddy Hield,Sacramento Kings,SG,1992-12-17


단일 값을 찾을 때에는 at과 iat이 loc나 iloc보다 빠르다.

In [144]:
#Austin Rivers에 대한 행 중에서 Birthday값을 추출한다.
nba.at["Austin Rivers", "Birthday"]

Timestamp('1992-08-01 00:00:00')

In [145]:
nba.iat[263, 1]

'PF'

## 4.8 Extracting Values from Series

In [152]:
#Salary Series에서 Damian Lillard에 대한 값을 찾는다.
nba["Salary"].loc["Damian Lillard"]

np.int64(29802321)

In [153]:
nba["Salary"].at["Damian Lillard"]

np.int64(29802321)

In [154]:
#Salary Series에서 234번째 행에 대한 값을 찾는다.
nba["Salary"].iloc[234]

np.int64(2033160)

In [155]:
nba["Salary"].iat[234]

np.int64(2033160)

## 4.9 Renaming Columns or Rows

In [156]:
nba.columns

Index(['Team', 'Position', 'Birthday', 'Salary'], dtype='object')

In [157]:
#nba의 column의 이름 중 Salary를 Pay로 변경한다.
nba.columns = ["Team", "Position", "Date of Birth", "Pay"]
nba.head(1)

,Team,Position,Date of Birth,Pay
Name,,,,
Shake Milton,Philadelphia 76ers,SG,1996-09-26,1445697


In [159]:
#key가 기존 열의 이름이고, 값이 새 이름인 dictionary를 columns매개변수에 전달하여 rename() method로 열의 이름을 변경한다.
#이때는 Date of Birth를 Birthday로 변경한다.
nba.rename(columns = { "Date of Birth": "Birthday" })

,Team,Position,Birthday,Pay
Name,,,,
Shake Milton,Philadelphia 76ers,SG,1996-09-26,1445697
Christian Wood,Detroit Pistons,PF,1995-09-27,1645357
PJ Washington,Charlotte Hornets,PF,1998-08-23,3831840
Derrick Rose,Detroit Pistons,PG,1988-10-04,7317074
Marial Shayok,Philadelphia 76ers,G,1995-07-26,79568
...,...,...,...,...
Austin Rivers,Houston Rockets,PG,1992-08-01,2174310
Harry Giles,Sacramento Kings,PF,1998-04-22,2578800
Robin Lopez,Milwaukee Bucks,C,1988-04-01,4767000


In [160]:
nba = nba.rename(columns = { "Date of Birth": "Birthday" })

In [161]:
nba.loc["Giannis Antetokounmpo"]

Team            Milwaukee Bucks
Position                     PF
Birthday    1994-12-06 00:00:00
Pay                    25842697
Name: Giannis Antetokounmpo, dtype: object

In [163]:
#index label의 값도 변경할 수 있다.
#Giannis Antetokounmpo라는 사람의 이름을 Greek Freak로 변경할 수 있다.

nba = nba.rename(
    index = { "Giannis Antetokounmpo": "Greek Freak" }    
)

In [164]:
nba.loc["Greek Freak"]

Team            Milwaukee Bucks
Position                     PF
Birthday    1994-12-06 00:00:00
Pay                    25842697
Name: Greek Freak, dtype: object

## 4.10 Resetting an Index

In [165]:
nba.set_index("Team").head()

,Position,Birthday,Pay
Team,,,
Philadelphia 76ers,SG,1996-09-26,1445697
Detroit Pistons,PF,1995-09-27,1645357
Charlotte Hornets,PF,1998-08-23,3831840
Detroit Pistons,PG,1988-10-04,7317074
Philadelphia 76ers,G,1995-07-26,79568


index label을 다른 column으로 변경하고 싶을 때 정보의 손실없이 변경하는 방법은 다음과 같다.

In [166]:
#일단 index label을 설정하지 않은 상태로 되돌린다.
nba.reset_index().head()

,Name,Team,Position,Birthday,Pay
0,Shake Milton,Philadelphia 76ers,SG,1996-09-26,1445697
1,Christian Wood,Detroit Pistons,PF,1995-09-27,1645357
2,PJ Washington,Charlotte Hornets,PF,1998-08-23,3831840
3,Derrick Rose,Detroit Pistons,PG,1988-10-04,7317074
4,Marial Shayok,Philadelphia 76ers,G,1995-07-26,79568


In [167]:
#그리고 index label을 설정한다.
nba.reset_index().set_index("Team").head()

,Name,Position,Birthday,Pay
Team,,,,
Philadelphia 76ers,Shake Milton,SG,1996-09-26,1445697
Detroit Pistons,Christian Wood,PF,1995-09-27,1645357
Charlotte Hornets,PJ Washington,PF,1998-08-23,3831840
Detroit Pistons,Derrick Rose,PG,1988-10-04,7317074
Philadelphia 76ers,Marial Shayok,G,1995-07-26,79568


In [168]:
nba = nba.reset_index().set_index("Team")

## 4.11 Coding Challenge

### 4.11.1 Problems

### 4.11.2 Solutions

In [106]:
nfl = pd.read_csv("nfl.csv", parse_dates = ["Birthday"])
nfl

,Name,Team,Position,Birthday,Salary
0,Tremon Smith,Philadelphia Eagles,RB,1996-07-20,570000
1,Shawn Williams,Cincinnati Bengals,SS,1991-05-13,3500000
2,Adam Butler,New England Patriots,DT,1994-04-12,645000
3,Derek Wolfe,Denver Broncos,DE,1990-02-24,8000000
4,Jake Ryan,Jacksonville Jaguars,OLB,1992-02-27,1000000
...,...,...,...,...,...
1650,Bashaud Breeland,Kansas City Chiefs,CB,1992-01-30,805000
1651,Craig James,Philadelphia Eagles,CB,1996-04-29,570000
1652,Jonotthan Harrison,New York Jets,C,1991-08-25,1500000
1653,Chuma Edoga,New York Jets,OT,1997-05-25,495000


In [107]:
nfl = nfl.set_index("Name")

In [108]:
nfl = pd.read_csv("nfl.csv", index_col = "Name", parse_dates = ["Birthday"])

In [109]:
nfl.head()

,Team,Position,Birthday,Salary
Name,,,,
Tremon Smith,Philadelphia Eagles,RB,1996-07-20,570000
Shawn Williams,Cincinnati Bengals,SS,1991-05-13,3500000
Adam Butler,New England Patriots,DT,1994-04-12,645000
Derek Wolfe,Denver Broncos,DE,1990-02-24,8000000
Jake Ryan,Jacksonville Jaguars,OLB,1992-02-27,1000000


In [110]:
# The two lines below are equivalent
nfl.Team.value_counts().head()
nfl["Team"].value_counts().head()

New York Jets          58
Kansas City Chiefs     56
Washington Redskins    56
New Orleans Saints     55
San Francisco 49Ers    55
Name: Team, dtype: int64

In [111]:
nfl.sort_values("Salary", ascending = False).head()

,Team,Position,Birthday,Salary
Name,,,,
Kirk Cousins,Minnesota Vikings,QB,1988-08-19,27500000
Jameis Winston,Tampa Bay Buccaneers,QB,1994-01-06,20922000
Marcus Mariota,Tennessee Titans,QB,1993-10-30,20922000
Derek Carr,Oakland Raiders,QB,1991-03-28,19900000
Jimmy Garoppolo,San Francisco 49Ers,QB,1991-11-02,17200000


In [112]:
nfl.sort_values(
    by = ["Team", "Salary"],
    ascending = [True, False]
)

,Team,Position,Birthday,Salary
Name,,,,
Chandler Jones,Arizona Cardinals,OLB,1990-02-27,16500000
Patrick Peterson,Arizona Cardinals,CB,1990-07-11,11000000
Larry Fitzgerald,Arizona Cardinals,WR,1983-08-31,11000000
David Johnson,Arizona Cardinals,RB,1991-12-16,5700000
Justin Pugh,Arizona Cardinals,G,1990-08-15,5000000
...,...,...,...,...
Ross Pierschbacher,Washington Redskins,C,1995-05-05,495000
Kelvin Harmon,Washington Redskins,WR,1996-12-15,495000
Wes Martin,Washington Redskins,G,1996-05-09,495000


In [113]:
nfl = nfl.reset_index().set_index(keys = "Team")
nfl.head(3)

,Name,Position,Birthday,Salary
Team,,,,
Philadelphia Eagles,Tremon Smith,RB,1996-07-20,570000
Cincinnati Bengals,Shawn Williams,SS,1991-05-13,3500000
New England Patriots,Adam Butler,DT,1994-04-12,645000


In [114]:
nfl.loc["New York Jets"].head()

,Name,Position,Birthday,Salary
Team,,,,
New York Jets,Bronson Kaufusi,DE,1991-07-06,645000
New York Jets,Darryl Roberts,CB,1990-11-26,1000000
New York Jets,Jordan Willis,DE,1995-05-02,754750
New York Jets,Quinnen Williams,DE,1997-12-21,495000
New York Jets,Sam Ficken,K,1992-12-14,495000


In [115]:
nfl.loc["New York Jets"].sort_values("Birthday").head(1)

,Name,Position,Birthday,Salary
Team,,,,
New York Jets,Ryan Kalil,C,1985-03-29,2400000


## 4.12 Summary